File: DailyChallenges/CPP/LeetCode_134.cpp
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int ts = 0, ls = 0, st=0;
        for (int i=0; i<gas.size(); ++i){
            ts += gas[i] - cost[i];
            ls += gas[i] - cost[i];
            if (ls < 0){
                ls = 0;
                st=i+1;
            }
        }
        return (ts < 0) ? -1 : st;
    }
};

File: DailyChallenges/Scala/LeetCode_134.scala
object Solution {
    def canCompleteCircuit(gas: Array[Int], cost: Array[Int]): Int = {
        val t = gas.zip(cost).map(p => p._1 - p._2).zipWithIndex.foldLeft((0,0,0))(
            (t,c)=> if (t._2 + c._1 < 0) (t._1 + c._1, 0, c._2 +1) else (t._1 + c._1, t._2 + c._1, t._3))
        if (t._1<0) -1 else t._3
    }
}
